### ISSUE TITLE
[docs] How to listen / bind to events

### ISSUE THREAD

**Issue by smeijer** [slothking-online/diagram#72]
I can see it's being done on https://app.graphqleditor.com, but I can't see to figure out how it works / where the hooks are.

Having `this.diagram = new Diagram()`, how can I add trigger that's being called when the diagram is changed through the canvas? (to call the serialize method). And how can I inject serialized data, when I would have changed the code manually?

**Comment by aexol** [comment]
```ts
class GraphController{
  serialise = ({ nodes, links }: { nodes: Node[]; links: Link[] }) => {
    this.nodes = nodes;
    this.links = links;
    const graphQLSchema = NodesToTree.parse(nodes, links);
    this.onSerialize && this.onSerialize(graphQLSchema);
    this.passSchema &&
      this.passSchema(
        {
          [ParsingFunction.graphql]: () => graphQLSchema,
          [ParsingFunction.typescript]: () =>
            TreeToTS.resolveTree(this.parser.parse(graphQLSchema)),
          [ParsingFunction.faker]: () => TreeToFaker.resolveTree(this.parser.parse(graphQLSchema))
        }[this.currentParsingFunction]()
      );
  };
  loadDefinitions = () => {
    this.definitions = Definitions.generate();
    this.diagram!.setDefinitions(this.definitions);
    this.diagram!.setSerialisationFunction(this.serialise);
  };
}
```
It is a part but you can see whats going on

**Comment by aexol** [comment]
tyou have setSerialisationFunction

**Comment by aexol** [comment]
@smeijer

**Comment by smeijer** [comment]
Thanks, turned out to be very simple indeed. `this.diagram.setSerialisationFunction` is basically an `onChange` that is called whenever a node is added/removed, or when the links change.

Worth mentioning: it's not being called when a node is being moved.

**Comment by smeijer** [comment]
@aexol, is there also an "selected" event available? So that we can get the `definition` of the currently selected node? And listen to the selection changes?

**Comment by aexol** [comment]
@smeijer no, but I can add that today later or tomorrow. If you want more events being handled I can add public event bus resolver to the project. Everything is event driven in this library so you can capture all actions. Sounds ok?

**Comment by aexol** [comment]
One more thing. Do you want to have pub/sub methods public or get one onEvent method on whole diagram

**Comment by aexol** [comment]
@smeijer

**Comment by aexol** [comment]
ok done @smeijer

**Comment by aexol** [comment]
```
this.diagram.setEventListener

**Comment by hzub** [comment]
@aexol @smeijer 
I't rather add possibility to register multiple listeners (i.e. addEventListener instead of setEventListener), also with an option to unregister those. I can do it later if you want.

**Comment by smeijer** [comment]
@aexol,

I'm not sure what you mean by

> One more thing. Do you want to have pub/sub methods public or get one onEvent method on whole diagram

I think what @hzub describes would be most usefull.

```js
diagram.on('select', (nodes) => {
  // nodes can be empty when clearing selection
  console.log('the following nodes have been selected', nodes); 
});

diagram.on('change', (nodes) => {
  // basically all changes to `nodes` and `links`
});
```

I'll definitely take a look at `setEventListener`.

**Comment by smeijer** [comment]
I've been playing with it, and although it's very interesting. I believe that `setEventListener` is too low level for most users.

It does have the `PickRequested` for example, but it's missing the `node selected`. Also, this stream requires me to filter for interesting events myself. While as an end user, I would just want to 'subscribe' to specific events instead.

**Comment by aexol** [comment]
DiagramEvents.NodeSelected is the event however it doesnt contain node now. I'll change that tommorow

**Comment by aexol** [comment]
And will change to the pub/sub model

**Comment by aexol** [comment]
Thank you for your feedback!

**Comment by smeijer** [comment]
You're welcome. I hope you don't see my issues as something negative. (I've created quite a few by now). It's my way of saying thanx :)

**Comment by hzub** [comment]
@aexol @smeijer 
What's the final conclusion here? Which events should be exposed to external pubsub interface?

**Comment by aexol** [comment]
@smeijer Could you help us with the solution of the problem? I mean what function should we provide to make event handling easy on your side?

**Comment by hzub** [comment]
@aexol i think that it's not a big deal to expose the API, but it'd be good to get an advice which of internal diagram's mechanisms should be exposed in such pattern.

**Comment by smeijer** [comment]
My needs are basically described in the comment above: https://github.com/slothking-online/diagram/issues/72#issuecomment-467898876

As a user, the need is because:

1) When selecting a node, I wish to show a form in an sidebar, matching that node. This, to edit settings like a name, or some data belonging to that node. Function arguments so to say.

```bash
# bill
[ subtotal ]  ---->  [ add taxes ] ----> [ send invoice ]
```

When I select `[ add taxes ]`, I might wish to show a form, that allows me to configure the tax percentage (`21%`).

Easiest way for me would be to have an `onSelect` hook, which is invoked when a selection has been made, or been cleared. For me, it doesn't matter if this is a `diagram.onSelect(...)`, or `diagram.on('select', () => {})` handler. Although, I do believe the second is used the most.

Second requirement I have, is to know when anything in my datamodel has been changed. Like I wrote before, this can currently be achieved by a combination of  `diagram.setSerialisationFunction(handleChange)` and `diagram.setPositionSerialisationFunction(handleChange)`, but that wasn't really obvious.

A handler like `diagram.on('change', ({ nodes, links, view }) => {})`, would be helpfull.

Use case, whenever I pan or zoom the canvas, or add/remove nodes or links, or move them arround, or rename them. I need to serialize and save to the database.

Basically, I need this `change` to be called, whenever my datamodel would change. Allowing me to save the datamodel, and close the browser. Whenever I open the page again, I could deserialize whatever I saved in the `onChange` handler, and resume where I left off.

**Comment by hzub** [comment]
@smeijer Oh, ok. Sorry that I didn't notice it before.
Ok, I guess that it won't be difficult - I'll try to tackle it asap.

**Comment by hzub** [comment]
@smeijer @aexol 
Mentioned changes are in PR #82 . From now on, you can
- bind (by `on()`) to `NodeSelected` event, which solves your first problem.
- bind to `DataModelChanged` and `ViewModelChanged` events, which propagate events about node and UI state, respectively. This solved second problem.
